iotilesensorgraph/iotile/sg/graph.py

```diff
@@ -514,3 +529,41 @@ def dump_streamers(self):
         """Dump all of the streamers in this sensor graph as a list of strings."""
 
         return [str(streamer) for streamer in self.streamers]
+
+    def dump_constant_database(self):
+        """Dump all of the constants in this sensor graph as a list of strings."""
+
+        return ["{}: {}".format(constant, val) for constant, val in self.constant_database.items()]
+
+    def dump_metadata_database(self):
+        """Dump all of the metadata in this sensor graph as a list of strings."""
+
+        return ["{}: {}".format(metadata, val) for metadata, val in self.metadata_database.items()]
+
+    def dump_config_database(self):
+        """Dump all of the config variables in this sensor graph as a list of strings."""
+
+        return ["{}: {}".format(config, val) for config, val in self.config_database.items()]
+
+    def calculate_sensorgraph_crc(self, polynomial=0x104C11DB7):
+        """Returns the calculated CRC of the sensor graph"""
+
+        sensor_graph_dump = self.dump_nodes() + self.dump_streamers() +\
+                            self.dump_constant_database() +\
+                            self.dump_metadata_database() +\
+                            self.dump_config_database()
+
+        sensor_graph_string = ', '.join(sensor_graph_dump)
+        sensor_graph_bytes = bytes(sensor_graph_string, 'utf-8')
```

Comment:
While this is the way to calculate the checksum of a file, the bytes representation of the 'utf-8' characters will not be available in a POD. There must be another representation that can be used for this such that a POD can be verified without the SGF file.

Reply:
I think the objective/goal is getting a bit convoluted here. Sounds like we are on a different page.

I thought that the intention with the config variable was to hold the calculated checksum of the `.sgf` file, and that the variable should be basically read only when a programmer wants to see that information so that they can compare it with an `.sgf` they have.

Are you saying with your statement that the firmware is now responsible for dumping out all of its sensor graph's content, so that the sensor graph python proxy can reconstruct the `SensorGraph` instance? or at least spit out a binary representation?

Lets try this exercise... I want to try to understand the use case that you are expecting from this feature. How do you see yourself or other programmers verify a device's sensor graph contents? What are the commands you would execute?
`iotile hw --port bled112 connect <uuid> controller sensor_graph calculate_checksum`
Where `calculate_checksum` returns a checksum after sending RPC's to the device to dump its sensor graph? And from there the user can compare that calculated checksum with the actual `.sgf`?

What I had in mind was just looking at the config variable and just trusting it for what it is.
`iotile hw --port bled112 connect <uuid> controller config_manager get_variable 0x9999`
And the user would just compare that value with the actual `.sgf`, or whatever list of checksums we have that we've associated with certain sensorgraphs.

Reply:
I think we need to first talk about what we want a binary representation of.

If its just the sensorgraph nodes, then [it already exists](https://code.eng.archsys.io/github.com/iotile/pod_controller@6d77dfa6bac718ea73e4934d57d9ebbb8f51091d/-/blob/lib_controller/python/support/sensorgraph.py#L134:9).
If we want it to include constants, firmware will need to know how to dump `rsl` constants.
And if we want config variables, we will need it to call RPC's to the config manager, which is a different interface. And do we want to include metadata in the binary representation as well?